hathor/builder.py

```diff
@@ -36,12 +40,308 @@
+        settings = self._get_settings()
+        reactor = self._get_reactor()
+        pubsub = self._get_or_create_pubsub()
+
+        peer_id = self._get_peer_id()
+
+        soft_voided_tx_ids = set(settings.SOFT_VOIDED_TX_IDS)
+        consensus_algorithm = ConsensusAlgorithm(soft_voided_tx_ids, pubsub)
+
+        wallet = self._get_or_create_wallet()
+        tx_storage = self._get_or_create_tx_storage()
+        indexes = tx_storage.indexes
+        assert indexes is not None
+
+        if self._enable_address_index:
+            indexes.enable_address_index(pubsub)
+
+        if self._enable_tokens_index:
+            indexes.enable_tokens_index()
+
+        if self._enable_utxo_index:
+            indexes.enable_utxo_index()
+
+        kwargs: Dict[str, Any] = {}
+
+        if self._enable_sync_v1 is not None:
+            kwargs['enable_sync_v1'] = self._enable_sync_v1
+
+        if self._enable_sync_v2 is not None:
+            kwargs['enable_sync_v2'] = self._enable_sync_v2
+
+        if self._stratum_port is not None:
+            kwargs['stratum_port'] = self._stratum_port
+
+        if self._network is None:
+            raise TypeError('you must set a network')
+
+        if self._event_manager is not None:
+            kwargs['event_manager'] = self._event_manager
+
+        manager = HathorManager(
+            reactor,
+            pubsub=pubsub,
+            consensus_algorithm=consensus_algorithm,
+            peer_id=peer_id,
+            tx_storage=tx_storage,
+            network=self._network,
+            wallet=wallet,
+            rng=self._rng,
+            checkpoints=self._checkpoints,
+            capabilities=self._capabilities,
+            environment_info=get_environment_info(self._cmdline, peer_id.id),
+            **kwargs
+        )
+
+        self.artifacts = BuildArtifacts(
+            settings=settings,
+            rng=self._rng,
+            reactor=reactor,
+            manager=manager,
+            pubsub=pubsub,
+            consensus=consensus_algorithm,
+            tx_storage=tx_storage,
+            indexes=indexes,
+            wallet=wallet,
+            rocksdb_storage=self._rocksdb_storage,
+        )
+
+        return self.artifacts
+
+    def check_if_can_modify(self) -> None:
+        if self.artifacts is not None:
+            raise ValueError('cannot modify after build() is called')
+
+    def set_event_manager(self, event_manager: EventManager) -> None:
+        self.check_if_can_modify()
+        self._event_manager = event_manager
+
+    def set_rng(self, rng: Random) -> None:
+        self.check_if_can_modify()
+        self._rng = rng
+
+    def set_checkpoints(self, checkpoints: List[Checkpoint]) -> None:
+        self.check_if_can_modify()
+        self._checkpoints = checkpoints
+
+    def set_capabilities(self, capabilities: List[str]) -> None:
+        self.check_if_can_modify()
+        self._capabilities = capabilities
+
+    def set_peer_id(self, peer_id: PeerId) -> None:
```

Comment:
A common builder pattern is to `return self` on every config method, allowing for method chaining and a less verbose api, for example:

```python
artifacts = (
    Builder()
        .set_event_manager(event_manager)
        .set_rng(rng)
        .set_checkpoints(checkpoints)
        .set_capabilities(capabilities)
        .set_peer_id(peer_id)
)
```

What do you think?

Reply:
Nice catch! Thanks! Done!